`MatchError: null` in `ChannelExecutor.unwind`

Hello. I was running benchmarks for my program and encountered a strange error randomly:

```
Exception in thread "main" Exception in thread "zio-fiber-9207130" scala.MatchError: null
        at zio.stream.internal.ChannelExecutor.unwind$1(ChannelExecutor.scala:39)
        at zio.stream.internal.ChannelExecutor.popAllFinalizers(ChannelExecutor.scala:45)
        at zio.stream.internal.ChannelExecutor.close(ChannelExecutor.scala:92)
        at zio.stream.internal.ChannelExecutor.restorePipe(ChannelExecutor.scala:25)
        at zio.stream.internal.ChannelExecutor.$anonfun$run$13(ChannelExecutor.scala:194)
        at zio.stream.internal.ChannelExecutor.$anonfun$popAllFinalizers$1(ChannelExecutor.scala:41)
        at zio.ZIO.$anonfun$$times$greater$1(ZIO.scala:89)
        at <empty>.Demo.main.fa(Demo.scala:7)
        at <empty>.Demo.main.fa(Demo.scala:6)
        at <empty>.Demo.main.fa(Demo.scala:8)
        at <empty>.Demo.main(Demo.scala:12)
Caused by: scala.MatchError: null
        at zio.stream.internal.ChannelExecutor.unwind$1(ChannelExecutor.scala:39)
        at zio.stream.internal.ChannelExecutor.popAllFinalizers(ChannelExecutor.scala:45)
        at zio.stream.internal.ChannelExecutor.close(ChannelExecutor.scala:92)
        at zio.stream.internal.ChannelExecutor.restorePipe(ChannelExecutor.scala:25)
        at zio.stream.internal.ChannelExecutor.$anonfun$run$13(ChannelExecutor.scala:194)
        at zio.stream.internal.ChannelExecutor.$anonfun$popAllFinalizers$1(ChannelExecutor.scala:41)
        at zio.ZIO.$anonfun$$times$greater$1(ZIO.scala:89)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1218)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1218)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1011)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:1218)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:934)
        at zio.internal.FiberRuntime.evaluateEffect(FiberRuntime.scala:389)
        at zio.internal.FiberRuntime.evaluateMessageWhileSuspended(FiberRuntime.scala:512)
        at zio.internal.FiberRuntime.drainQueueOnCurrentThread(FiberRuntime.scala:229)
        at zio.internal.FiberRuntime.run(FiberRuntime.scala:138)
        at zio.internal.ZScheduler$$anon$4.run(ZScheduler.scala:476)
```

I can reproduce the problem in my environment with the following code:

```scala
import zio.stream.ZStream
import zio.Unsafe
import zio.ZIO

object Demo {
  def main(args: Array[String]): Unit = {
    val fa = ZStream.unwrapScoped {
      ZStream.range(0, 100).toQueue().map(q => ZStream.fromQueue(q).flattenTake)
    }.runDrain

    Unsafe.unsafe { implicit unsafe =>
      while (true) // Also reproduced by `fa.forever`.
        zio.Runtime.default.unsafe.run(fa).getOrThrowFiberFailure()
    }
  }
}
```
The time it takes for the error to occur is random. In my environment, it can be anywhere from a few minutes to a few dozen minutes.

ZIO: 2.0.5, 2.0.6
JVM: Temurin-11.0.17+8
Scala: 2.13.10

--- a/streams/shared/src/main/scala/zio/stream/ZChannel.scala
+++ b/streams/shared/src/main/scala/zio/stream/ZChannel.scala
@@ -1012,9 +1012,9 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
     ev1: Any <:< InElem,
     ev2: OutElem <:< Nothing,
     trace: Trace
-  ): ZIO[Env with Scope, OutErr, OutDone] =
-    ZIO
-      .acquireReleaseExit(
+  ): ZIO[Env with Scope, OutErr, OutDone] = {
+    def run(promise: Promise[OutErr, OutDone], scope: Scope) = ZIO
+      .acquireReleaseExitWith(
         ZIO.succeed(
           new ChannelExecutor[Env, InErr, InElem, InDone, OutErr, OutElem, OutDone](
             () => self,
@@ -1022,12 +1022,11 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
             executeCloseLastSubstream = identity[URIO[Env, Any]]
           )
         )
-      ) { (exec, exit) =>
+      ) { (exec, exit: Exit[OutErr, OutDone]) =>
         val finalize = exec.close(exit)
-        if (finalize ne null) finalize
+        if (finalize ne null) finalize.tapErrorCause(cause => scope.addFinalizer(ZIO.refailCause(cause)))
         else ZIO.unit
-      }
-      .flatMap { exec =>
+      } { exec =>
         ZIO.suspendSucceed {
           def interpret(channelState: ChannelExecutor.ChannelState[Env, OutErr]): ZIO[Env, OutErr, OutDone] =
             channelState match {
@@ -1047,8 +1046,19 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
             }
 
           interpret(exec.run().asInstanceOf[ChannelState[Env, OutErr]])
+            .intoPromise(promise) *> promise.await <* ZIO.never
         }
       }
+
+    for {
+      parent  <- ZIO.scope
+      child   <- parent.fork
+      promise <- Promise.make[OutErr, OutDone]
+      fiber   <- run(promise, child).forkScoped
+      done    <- promise.await
+      _       <- fiber.inheritAll
+    } yield done
+  }
 
   /**
    * Run the channel until it finishes with a done value or fails with an error.

